plenum/server/primary_selector.py

```diff
@@ -89,3 +92,60 @@ def next_primary_replica_name_for_backup(self, instance_id, master_primary_rank,
     def start_election_for_instance(self, instance_id):
         raise NotImplementedError("Election can be started for "
                                   "all instances only")
+
+    def on_catchup_complete(self):
+        # Select primaries after usual catchup (not view change)
+        ledger = self.node.getLedger(AUDIT_LEDGER_ID)
+        if len(ledger) == 0:
+            if self.viewNo != 0:
+                raise LogicError('If audit ledger is empty, view_no must be 0. '
+                                 'Because this node just started, did not order any txn '
+                                 'and did not make a view change.')
+            for replica in self.replicas.values():
+                if replica.primaryName is not None:
+                    raise LogicError('If audit ledger is empty, '
+                                     'all primaries must not be set yet')
+            if len(self.replicas) != self.node.requiredNumberOfInstances:
+                raise LogicError('If audit ledger is empty, all replicas'
+                                 'must be active')
+            self.node.select_primaries()
+        else:
+            self.node.backup_instance_faulty_processor.restore_replicas()
+            self.node.drop_primaries()
+            self.node.viewNo = get_payload_data(ledger.get_last_committed_txn())[AUDIT_TXN_VIEW_NO]
+            self.node.primaries = self.get_last_audited_primaries()
+            if len(self.replicas) != len(self.node.primaries):
+                raise LogicError('Audit ledger has inconsistent number of nodes')
+            if any(p not in self.node.nodeReg for p in self.node.primaries):
+                raise LogicError('Audit ledger has inconsistent names of nodes')
+            # Similar functionality to select_primaries
+            for instance_id, replica in self.replicas.items():
+                if instance_id == 0:
+                    self.node.start_participating()
+                replica.primaryChanged(self.node.primaries[instance_id])
+                self.node.primary_selected(instance_id)
+
+            # Primary propagation
```

Comment:
Should we do this regardless of the way how primaries are restored, that is move it from outside else condition?

Reply:
yes, moved